R/fit_baselines_one_location.R

```diff
@@ -0,0 +1,158 @@
+  horizons_to_forecast <- 1:max(effective_horizons)
+  h_adjustments <- min(effective_horizons) - 1
+
+  if (h_adjustments > 0) {
+    cli::cli_warn(
+      "forecasts requested for a time index beyond the provided {.arg target_ts},
+        replacing missing target data with {.val {h_adjustments}} target observations"
+    )
+    target_ts <- target_ts |>
+      dplyr::group_by(dplyr::across(c("location"))) |>
+      tidyr::complete(
+        time_index = unique(c(target_ts$time_index, last_data_date + h_adjustments * ts_temp_res)),
+        fill = list(observation = target_ts$observation[target_ts$time_index == last_data_date]),
+      )
+
+    last_data_date <- max(target_ts$time_index)
+    effective_horizons <- as.integer(actual_target_dates - last_data_date) / ts_temp_res
+    horizons_to_forecast <- 1:max(effective_horizons)
+    h_adjustments <- min(effective_horizons) - 1
+  }
+
+  # get predictions for all model_variations
+  predictions <- purrr::pmap_dfr( #tibble, each 1x1 row contains predictions for 1 model
+    model_variations,
+    get_baseline_predictions,
+    target_ts = target_ts,
+    effective_horizons = horizons_to_forecast,
+    origin = ifelse(temporal_resolution == "weekly", "obs", "median"),
+    n_sim = 100000,
+    quantile_levels = quantile_levels,
+    n_samples = n_samples,
+    round_predictions = round_predictions,
+    seed = seed
+  )
+
+  # extract forecasts
+  extracted_outputs <-
+    dplyr::bind_cols(model_variations, predictions) |>
+    tidyr::unnest(cols = "forecasts") |>
+    dplyr::mutate(
+      location = unique(target_ts$location),
+      horizon = .data[["horizon"]],
+      target_end_date = last_data_date + ts_temp_res * .data[["horizon"]],
+      .before = "horizon"
+    )
+
+  # h_adjustments = 0 (min(actual_target_dates) occurs period after last observed value)
+
+  if (h_adjustments == 0) {
+    model_outputs <- extracted_outputs |>
+      dplyr::mutate(
+        reference_date = reference_date,
+        horizon = as.numeric((.data[["target_end_date"]] - as.Date(reference_date)) / ts_temp_res),
+        .before = "horizon"
+      )
+  } else if (h_adjustments < 0) {
+    model_outputs <- extracted_outputs |>
+      dplyr::group_by(dplyr::across(c("transformation", "symmetrize", "window_size", "location", "output_type"))) |>
+      tidyr::complete(target_end_date = actual_target_dates, .data[["output_type_id"]]) |>
+      dplyr::ungroup() |>
+      dplyr::mutate(
+        reference_date = reference_date,
+        horizon = as.numeric((.data[["target_end_date"]] - as.Date(reference_date)) / ts_temp_res),
+        .before = "horizon"
+      ) |>
+      dplyr::left_join(
+        dplyr::filter(target_ts, .data[["time_index"]] %in% actual_target_dates),
+        by = c("location", "target_end_date" = "time_index")
+      ) |>
+      dplyr::mutate(
+        value = ifelse(is.na(.data[["value"]]), .data[["observation"]], .data[["value"]])
```

Comment:
I would have expected this `ifelse` to be organized the other way around: `value = ifelse(is.na(.data[["observation"]]), .data[["value"]], .data[["observation"]])`.  This would say, "if we didn't just merge in an observation for this target date, keep the predicted value; otherwise, use that observation."  What am I missing?

Reply:
This ifelse statement has the logic "if we didn't predict a value for this target date, use the observed value instead". Basically, I wrote it to avoid using extra computational power to calculate forecasts that are ultimately going to be replaced by the observed value, which happens when at least some of our predictions are for target end dates that have already occurred.

Again, this is something that the original trends ensemble functions did, which I have chosen to preserve, but I'm open to changing the implementation, removing the functionality, etc if we don't like what's there currently.

Reply:
Thanks, I understand now.  It makes sense to set things up so as to avoid any unnecessary computation!